drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/DMNScenarioValidation.java

```diff
@@ -0,0 +1,126 @@
+     * - field type changed
+     * @param simulation
+     * @param kieContainer
+     * @return
+     */
+    @Override
+    public List<FactMappingValidationError> validate(Simulation simulation, KieContainer kieContainer) {
+        List<FactMappingValidationError> errors = new ArrayList<>();
+        String dmnFilePath = simulation.getSimulationDescriptor().getDmnFilePath();
+        DMNModel dmnModel = getDMNModel(kieContainer, dmnFilePath);
+
+        for (FactMapping factMapping : simulation.getSimulationDescriptor().getFactMappings()) {
+            if (isToSkip(factMapping)) {
+                continue;
+            }
+
+            String nodeName = factMapping.getFactIdentifier().getName();
+
+            DMNType rootDMNType;
+            try {
+                rootDMNType = dmnModel.getDecisionByName(nodeName) != null ?
+                        dmnModel.getDecisionByName(nodeName).getResultType() :
+                        dmnModel.getInputByName(nodeName).getType();
+            } catch (NullPointerException e) {
+                errors.add(createNodeChangedError(factMapping, "node not found"));
+                continue;
+            }
+
+            List<String> steps = expressionElementToString(factMapping);
+
+            // error if direct mapping (= simple type) but it is a composite
+            // NOTE: context is a special case so it is composite even if no fields are declared
+            Type rootType = getRootType((BaseDMNTypeImpl) rootDMNType);
+            if (!CONTEXT.equals(rootType) && steps.size() == 0 && rootDMNType.isComposite()) {
+                errors.add(createNodeChangedError(factMapping, rootDMNType.getName()));
+                continue;
+            }
+
+            try {
+                DMNType fieldType = navigateDMNType(rootDMNType, steps);
+
+                if (!isDMNFactMappingValid(factMapping.getClassName(), factMapping, fieldType)) {
+                    errors.add(createFieldChangedError(factMapping, fieldType.getName()));
+                }
+            } catch (IllegalStateException e) {
+                errors.add(createGenericError(factMapping, e.getMessage()));
+            }
+        }
+        return errors;
+    }
+
+    private boolean isDMNFactMappingValid(String typeName, FactMapping factMapping, DMNType dmnType) {
+        // NOTE: Any/Undefined is a special case where collection is true
+        Type rootType = getRootType((BaseDMNTypeImpl) dmnType);
+        boolean isCoherent = UNKNOWN.equals(rootType) || ScenarioSimulationSharedUtils.isList(typeName) ==
+                dmnType.isCollection();
+        if (!isCoherent) {
+            return false;
+        }
+        String factMappingType = ScenarioSimulationSharedUtils.isList(typeName) ?
+                factMapping.getGenericTypes().get(0) :
+                typeName;
+
+        if (Objects.equals(factMappingType, dmnType.getName())) {
+            return true;
+        }
+        return false;
```

Comment:
I think [1] is same as [2]

[1]
```java
if (Objects.equals(factMappingType, dmnType.getName())) {
            return true;
        }
        return false;
```

[2]
```java
return Objects.equals(factMappingType, dmnType.getName());
```

Reply:
Yes, you are right, done :)